qiskit/transpiler/passes/extension_mapper/extension_mapper.py

```diff
@@ -0,0 +1,197 @@
+placement of quantum registers onto the architecture _only_. This does not include a circuit.
+We then use an algorithm that, given such a mapping, finds a 4-approximation of the best swap
+circuit (in size). Our compiler then takes such a mapping, computes the swaps using the routing
+algorithm, and finally places the mapping circuit followed by as many gates as possible (without
+remapping).
+
+Our starting point for the mapper is the qiskit swap mapper.
+We rewrote it to be specifically focused for optimizing in size.
+Simplifying it a lot along the way.
+We, however, constrain the qiskit mapper quite a bit, by allowing it to try only |V|^2/8 swaps.
+When it fails we fall back to our own code, the extension mapper. The extension mapper tries to
+first place qubits onto nodes such that the number of swaps is minimized.
+Then it tries to extend this placement with more qubits for which it holds that
+it is cheaper to place them now than in the next iteration.
+Finally, we compute if this procedure has provided a nice placement.
+If not, we simply place just the cheapest gate and go to the next iteration.
+
+- Did you use any previously published schemes? Cite relevant papers.
+* Approximation and Hardness for Token Swapping by Miltzow et al. (2016)
+    ArXiV: https://arxiv.org/abs/1602.05150
+    It describes the 4-approximation algorithm. We then generalized it slightly to support partial
+    token assignments.
+
+- What packages did you use in your code and for what part of the algorithm?
+The packages are listed in the Pipfile(.lock)
+qiskit and networkx are the main packages.
+Qiskit for the DAGCircuit data structure.
+NetworkX for graph operations.
+
+- How general is your approach? Does it work for arbitrary coupling layouts (qubit number)?
+Yes it does. If the layout is connected.
+
+- Are there known situations when the algorithm fails?
+When running you may encounter https://github.com/QISKit/qiskit-sdk-py/issues/239
+
+Future work:
+* Tweak the limit for the qiskit mapper from |V|^2/8 to something better.
+* Compute both extension mapper and qiskit mapper circuits and pick the 'best' one.
+
+"""
+
+import re
+import copy
+import pprint
+import logging
+
+from qiskit import QuantumRegister
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler._basepasses import TransformationPass
+from qiskit.mapper import cx_cancellation, optimize_1q_gates
+
+from .src import permutation
+
+from .src.mapping import size
+from .src.compiler import compile_to_arch
+from .src.mapping.util import direction_mapper
+from .src.mapping.placement import Placement
+from .src.permutation.general import ApproximateTokenSwapper
+
+logger = logging.getLogger(__name__)
+
+
+class ExtensionMapper(TransformationPass):
+    """Extension mapper implemented as a transpiler pass."""
+    def __init__(self, coupling_map, desired_layout=None):
+        self._coupling_map = coupling_map
+        self._desired_layout = desired_layout
```

Comment:
Does the `_desired_layout` work? It seems like it's just ignored here. Is there anyway to tell the mapper that I want `q[0]` mapped to `q[5]` on the chip?
If it does work, it would be helpful to have a test for it like you have in the LookaheadMapper.

Reply:
Good catch, this was left over from a prior commit. Updated to remove `desired_layout` entirely in favor of #1299.